horovod/common/operations.cc

```diff
@@ -843,22 +936,37 @@ void PerformOperation(TensorTable& tensor_table, MPIResponse response) {
                                      cudaMemcpyDeviceToDevice, stream))
           offset += it->tensor->size();
         }
+
+#if HOROVOD_GPU_ALLREDUCE != 'D'
         if (timeline.Initialized()) {
```

Comment:
Why do you side-step check for `timeline.Initialized()`?

Reply:
The DDL allreduce is a synchronous operation and uses its own Cuda streams. So we need to record an event on Horovod stream and wait for it before calling DDL to make all the work is done up to this point on the Horovod stream.
That is also why i need to start the finalizer_thread before all the events get added to the Horovod stream otherwise your timeline would be messed up because ddl is syncing the operation in the middle of all the events. Does that make sense?

Reply:
I see - makes sense.  I actually do some of the things along the similar lines in #329.  I think it makes sense to merge in refactorings from #329 first, and then your code change will actually be much simpler.

Reply:
ok will wait on #329.

Reply:
@alsrgv How is your PR #329 going?

Reply:
Waiting to hear back from a user in #312 to determine whether it works and whether it's faster.

Reply:
@nvcastet, the #329 has been pushed.  Please rebase/simplify this PR.